rust/kernel/print.rs

```diff
@@ -10,7 +10,69 @@ use core::cmp;
 use core::fmt;
 
 use crate::bindings;
-use crate::c_types::c_int;
+use crate::c_types::{c_char, c_int, c_void};
+
+// Called from vsprintf with format specifier %pfr.
+#[no_mangle]
+unsafe fn rust_fmt_argument(
+    buf: *mut c_char,
+    end: *mut c_char,
+    ptr: *const c_void,
+    limit: c_int,
+) -> *mut c_char {
+    use fmt::Write;
+
+    // Use `usize` to use saturating_* functions.
+    struct Writer {
+        buf: usize,
+        end: usize,
+        limit: usize,
+    }
+
+    impl Write for Writer {
+        fn write_str(&mut self, s: &str) -> fmt::Result {
+            // Maximum possible length to write, bounded by limit.
+            let len = cmp::min(s.len(), self.limit);
+
+            // `buf` value after writing `len` bytes. This does not have to be bounded
+            // by `end`, but we don't want it to wrap around to 0.
+            let buf_new = self.buf.saturating_add(len);
+
+            // Amount that we can copy. `saturating_sub` ensures we get 0 if
+            // `buf` goes past `end`.
+            let len_to_copy = cmp::min(buf_new, self.end).saturating_sub(self.buf);
+
+            // SAFETY: In any case, `buf` is non-null and properly aligned.
+            // If `len_to_copy` is non-zero, then we know `buf` has not past
+            // `end` yet and so is valid.
+            unsafe {
+                core::ptr::copy_nonoverlapping(
+                    s.as_bytes().as_ptr(),
+                    self.buf as *mut u8,
+                    len_to_copy,
+                )
+            };
+
+            self.buf = buf_new;
+            if s.len() <= self.limit {
+                self.limit -= s.len();
+                Ok(())
```

Comment:
I think it would be clearer for this to be
```rust
if let Some(new_limit) = self.limit.checked_sub(s.len()) {
    self.limit = new_limit;
    Ok(())
} else {
```

Reply:
How about
```rust
self.limit = self.limit.checked_sub(s.len()).ok_or(fmt::Error)?;
Ok(())
```

Reply:
Yeah, that seems even better!